### PR TITLE
ESQL: Load values a different way

### PR DESCRIPTION
**Pull request by nik9000**
This changes how we load values in ESQL, delegating to the `MappedFieldType` like we do with doc values and synthetic source. This allows a much more OO way of getting the loads working which makes that path much easier to read. And! It means those code paths look like doc values. So there's symmetry. It's like it rhymes.

There are a few side effects here:
1. It's fairly simple to load from ordinals efficiently. I wrote some block-at-a-time code for resolving ordinals and it's about twice as fast. With more work it should be possible to make custom ordinal-shaped blocks move through the system to save space and speed things up.
2. Most fields can now be loaded from `_source`. Everything that can be loaded from `_source` in scripts will load from `_source` in ESQL.
3. We get a *lot* more tests for loading fields in different configurations by piggybacking on the synthetic source testing framework. 
4. Loading from `_source` no longer sorts the fields. Same for stored fields. Now we keep them in whatever they were stored in. This is a pretty marginal time save because loading from `_source` is so much more time consuming than the sort. But it's something.

### CODE REVIEW

**Comment by nik9000** [comment]
Here's the speedup from decoding ordinals in a more sensible way:
```
Before ValuesSourceReaderBenchmark.benchmark  in_order  keyword  avgt    7  369.033 ± 12.640  ns/op
 After ValuesSourceReaderBenchmark.benchmark  in_order  keyword  avgt    7  142.628 ±  2.022  ns/op
```

One thing that's crying out to me - we really want to load from stored fields like `_source` in a different way. Right now we load just like doc values, but those are fundamentally row-at-a-time mechanisms. And we can do that. Just not right now.

File: x-pack/plugin/esql/qa/server/single-node/src/yamlRestTest/resources/rest-api-spec/test/90_non_indexed.yml [hunk 1]
```diff
@@ -130,18 +130,18 @@ unsupported:
...
+  - match: { values.0.9: 10 }
   - match: { values.0.10: "192.168.0.1" }
-  - match: { values.0.11: null }
+  - match: { values.0.11: null}
   - match: { values.0.12: "foo" }
-  - match: { values.0.13: "foo" } # this is a special case, ESQL can retrieve keywords from source
+  - match: { values.0.13: "foo" }
   - match: { values.0.14: 20 }
-  - match: { values.0.15: null }
+  - match: { values.0.15: 20 }
```
Comment (nik9000) [inline x-pack/plugin/esql/qa/server/single-node/src/yamlRestTest/resources/rest-api-spec/test/90_non_indexed.yml:147]: These are mostly supported now. I suppose they should move out.

**Comment by nik9000** [comment]
image

**Comment by dnhatn** [comment]
> Before ValuesSourceReaderBenchmark.benchmark  in_order  keyword  avgt    7  369.033 ± 12.640  ns/op
> After ValuesSourceReaderBenchmark.benchmark  in_order  keyword  avgt    7  142.628 ±  2.022  ns/op

Wow, this is awesome.

**Comment by nik9000** [comment]
> Wow, this is awesome.

Just use the ordinals properly and everything is faster!

Seriously, though, if we had a BytesRefBlock that allowed an extra layer of indirection we could resolve the ordinals one time and let them flow through the system. That'd be pretty sweet!

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java [hunk 2]
```diff
@@ -194,7 +194,7 @@ private void runComputeOnRemoteNodes(
         List<TargetNode> targetNodes,
         Supplier<ActionListener<DataNodeResponse>> listener
     ) {
-        // Do not complete the exchange sources until we have linked all remote sinks
+        // Do not complete the exchange factories until we have linked all remote sinks
```
Comment (dnhatn) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java:197]: inadvertent renaming

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java [hunk 3]
```diff
@@ -71,6 +71,7 @@ public TransportEsqlQueryAction(
         exchangeService.registerTransportHandler(transportService);
         this.exchangeService = exchangeService;
         this.enrichPolicyResolver = new EnrichPolicyResolver(clusterService, transportService, planExecutor.indexResolver());
+
```
Comment (dnhatn) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java:74]: leftover?

File: server/src/main/java/org/elasticsearch/index/mapper/BlockLoader.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        // TODO support non-singleton ords
+    }
+
+    /**
+     * A builder for typed values. For each document you may either call
+     * {@link #appendNull}, {@code append<Type>}, or
+     * {@link #beginPositionEntry} followed by two or more {@code append<Type>}
+     * calls, and then {@link #endPositionEntry}.
+     */
+    interface Builder {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/mapper/BlockLoader.java:133]: I think we should make this Builder Releasable and release it if we hit a breaker when reading values. Let's do this in a follow-up.

**Review by dnhatn — APPROVED** [review]
Beautiful! I think we can leverage the sequential stored-fields reader after this change too. Thank you, Nik!

**Comment by nik9000** [comment]
> I think we can leverage the sequential stored-fields reader after this change too.

I'd love to rework stored fields a bit before trying that. I think there should be a "load stored fields" operator. And we can do that by expanding `BlockLoader` to add stuff like `storedFieldsInfo` or something. We could, like, merge all the stored field loading. and if we did that we could super use the sequential reader. Or not. I mean, we have a list of doc ids we're loading then and we could use it!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeService.java [hunk 5]
```diff
@@ -42,9 +42,9 @@
 import java.util.concurrent.Executor;
 
 /**
- * {@link ExchangeService} is responsible for exchanging pages between exchange sinks and sources on the same or different nodes.
+ * {@link ExchangeService} is responsible for exchanging pages between exchange sinks and factories on the same or different nodes.
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeService.java:45]: That rename I did with IntelliJ had a huge blast radius!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-Block.java.st [hunk 6]
```diff
@@ -222,7 +223,7 @@ $endif$
         return blockFactory.newConstant$Type$BlockWith(value, positions);
     }
 
-    sealed interface Builder extends Block.Builder permits $Type$BlockBuilder {
+    sealed interface Builder extends Block.Builder, BlockLoader.$Type$Builder permits $Type$BlockBuilder {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-Block.java.st:226]: ++

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java (hunk 3 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java:74]: Yeah. that's a silly change. I can't quite remember how I got there either!

File: server/src/main/java/org/elasticsearch/index/mapper/BlockLoader.java (hunk 4 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/BlockLoader.java:133]: ++. I think we *have* to do it to properly track memory on blocks built by field loading. We just aren't doing that yet.

**Comment by nik9000** [comment]
I talked to @dnhatn and @ChrisHegarty about the funny `interface` I made in code that only has a production impl in `compute`. It's a "cute" trick to prevent having to drag all of the `Block` infrastructure into `server`. It's a *lot* to drag around and I think it'd be a big change to move it. So I made the interface. I've grown to like it a fair bit because there's lots of complexity in `compute` that I don't have to share with everyone. I really like that hacking on `Block` doesn't mean I have to recompile the world. I get that it's weird, but I've grown to like it.

The interface we expose is *fairly* small. Not actually small. But, like, small-ish. We may one day pull `Block` into server and remove the interface. But for now we're going to avoid it.

**Comment by nik9000** [comment]
This also significantly lowered the per-field overhead of loading fields. Especially empty fields. It turns out that *sometimes* this is a huge deal.

**Pull request merged**